schema/elements.go

```diff
@@ -27,6 +29,15 @@ type Schema struct {
 
 	once sync.Once
 	m    map[string]TypeDef
+
+	// Once used to protect the initialization of `lock` field.
+	lockOnce sync.Once
+	// Lock which protects writes to resolvedTypes. Used as pointer so that
+	// schema may be used as a value type
+	lock *sync.Mutex
```

Comment:
Do you need to make this a pointer? What bad thing happens if you omit the Once? If the problem is that it's bad to copy Mutexes, isn't it also bad to copy Onces?

Reply:
I trusted `go vet` too much it looks like here. Originally I had `sync.Mutex` as a value and `go vet` was throwing errors. 

I changed to this method since I saw `sync.Once` was already successfully used a few lines above in the original code, so I assumed its implementation would be safe. However, looking into the source for `Once`, I see it just stores a `Mutex` by value.

Now I am questioning whether the original code is safe. And wondering the correct approach: 

1. to use a `sync.Mutex` here and manually check if any copies are safe. Silencing go vet errors?
2. change usages of `Schema` to be as a pointer everywhere. From a brief review `Schema` is usually used as a pointer, but there are a few places where its usage as a value might be problematic:

https://github.com/kubernetes-sigs/structured-merge-diff/blob/daa3a199edcc48f138daf17ff9ec1ea875f563b8/typed/parser.go#L30-L33


(I am unsure of the consequences of changing the below to use pointers, it might be ok)
https://github.com/kubernetes-sigs/structured-merge-diff/blob/ded43451896507b64bce2978534fc6a534d3042e/schema/equals_test.go#L34-L39

https://github.com/kubernetes-sigs/structured-merge-diff/blob/f0a394c4104958b8b907bbe5a496cd55406f9d32/schema/elements.go#L39-L43

Reply:
We should probably change all uses to be a pointer, yeah...

Reply:
latest changes refactor usages of `Schema` to use a pointer